### PR TITLE
V1.4 - SSE upgrades and bug fixes

### PR DESCRIPTION
**Pull request by StrangeZak**

### CODE REVIEW

**Comment by StrangeZak** [comment]
Ive noticed in all of our vector "constructors"  we clear the vector to zero. Since we require the user to pass in all the needed elements to fill out the vector is there really any reason for us to zero it. I know it will really just get rid of ZeroMemory call, but still it just something on my mind. Any opinions on this @bvisness

**Comment by bvisness** [comment]
As far as I’m concerned we can just get rid of the zeroing. I think we
could get rid of the zeroing pretty much everywhere in Handmade Math.

**Comment by StrangeZak** [comment]
@bvisness Awesome, ill do it for v1.4.

**Comment by StrangeZak** [comment]
@bvisness Got rid of Zeroing. If you do a check over this that'd be great. Also, NormalizeZero test is failing i dont think we need this test anymore do we.

**Comment by StrangeZak** [comment]
Maybe i also just did something wrong when removing all the zeroing

**Comment by kiljacken** [comment]
@StrangeZak By removing the zeroing from the normalize functions we now return an uninitialized value when the length of the input vector is zero.

**Comment by StrangeZak** [comment]
@kiljacken Oh yeah, thats not good. This is what happens when you tired code. Will fix now

**Comment by StrangeZak** [comment]
I just added the zeroing back in im sure its not a big deal.

**Comment by StrangeZak** [comment]
Not sure if this is good enough for release now, or if i should SSE the vec2 and vec3s opinions anyone

**Comment by bvisness** [comment]
You might as well if you're in the mood?

**Comment by StrangeZak** [comment]
Ill try and hammer it out today.

**Comment by StrangeZak** [comment]
Not quiet sure on how my recent commit broke, A LOT of things.

**Comment by StrangeZak** [comment]
Maybe im just blind though

**Comment by bvisness** [comment]
@StrangeZak Only the SSE tests are failing. And because it appears that vector initialization is broken, lots of other tests broke too.

In particular, it seems that it's lines like these that are failing:

```cpp
            //
            // Test vec3
            //
            hmm_vec3 v3 = HMM_Vec3(1.0f, 2.0f, 3.0f);
            hmm_vec3 v3i = HMM_Vec3i(1, 2, 3);

            EXPECT_FLOAT_EQ(v3.X, 1.0f);
            EXPECT_FLOAT_EQ(v3.Y, 2.0f);
            EXPECT_FLOAT_EQ(v3.Z, 3.0f);
            EXPECT_FLOAT_EQ(v3.U, 1.0f);
            EXPECT_FLOAT_EQ(v3.V, 2.0f);
            EXPECT_FLOAT_EQ(v3.W, 3.0f);
            EXPECT_FLOAT_EQ(v3.R, 1.0f);
            EXPECT_FLOAT_EQ(v3.G, 2.0f);
            EXPECT_FLOAT_EQ(v3.B, 3.0f);
            EXPECT_FLOAT_EQ(v3.Elements[0], 1.0f);
            EXPECT_FLOAT_EQ(v3.Elements[1], 2.0f);
            EXPECT_FLOAT_EQ(v3.Elements[2], 3.0f);
            EXPECT_FLOAT_EQ(v3.XY.Elements[0], 1.0f);
            EXPECT_FLOAT_EQ(v3.XY.Elements[1], 2.0f);
            EXPECT_FLOAT_EQ(v3.YZ.Elements[0], 2.0f); // fail
            EXPECT_FLOAT_EQ(v3.YZ.Elements[1], 3.0f); // fail
            EXPECT_FLOAT_EQ(v3.UV.Elements[0], 1.0f);
            EXPECT_FLOAT_EQ(v3.UV.Elements[1], 2.0f);
            EXPECT_FLOAT_EQ(v3.VW.Elements[0], 2.0f); // fail
            EXPECT_FLOAT_EQ(v3.VW.Elements[1], 3.0f); // fail
```

**Comment by bvisness** [comment]
By the way, when you're looking at the test results, the numbers in the parentheses are the line number in hmm_test.c that failed. So this test...

```cpp
- [FAIL] (163) Expected 2.000000, got 0.000970
```

...failed on line 163.

**Comment by StrangeZak** [comment]
Yes, i understand that only the SSE tests are failing but with the previous commit everything was fine! Thats what im confused i didn't change much in the most recent commit

**Comment by StrangeZak** [comment]
All i did to vec3 was change around Cross, and add the     __m128 InternalElementsSSE; to the union

**Comment by StrangeZak** [comment]
@bvisness Reverted branch back one commit, good to merge to master?

File: HandmadeMath.h [hunk 1]
```diff
@@ -1380,13 +1451,15 @@ HMM_EqualsVec3(hmm_vec3 Left, hmm_vec3 Right)
...
-    hmm_bool Result = 0;
+    hmm_bool Result;
 
     Result = (Left.X == Right.X && Left.Y == Right.Y && Left.Z == Right.Z && Left.W == Right.W);
 
     return (Result);
 }
 
+// TODO(zak): Remove before pushing to main.
+// I dont think its worth us SSEing HMM_Mat4 since its just a mem-zero
```
Comment (bvisness) [inline HandmadeMath.h]: @StrangeZak We still have these comments here, even though it appears we're not doing any SSE in here.

File: HandmadeMath.h [hunk 2]
```diff
@@ -1395,11 +1468,14 @@ HMM_Mat4(void)
     return (Result);
 }
 
+// TODO(zak): Remove before pushing to main.
+// I dont think its worth us SSEing HMM_Mat4d since it would be the same ammount of instructions,
+// since we have to address each row individually.
```
Comment (bvisness) [inline HandmadeMath.h]: and here.

File: HandmadeMath.h [hunk 3]
```diff
@@ -1499,8 +1587,18 @@ HMM_MultiplyMat4(hmm_mat4 Left, hmm_mat4 Right)
...
 HMM_MultiplyMat4f(hmm_mat4 Matrix, float Scalar)
 {
-    hmm_mat4 Result = HMM_Mat4();
-
+    hmm_mat4 Result;
+    
+#ifdef HANDMADE_MATH__USE_SSE
+    // TODO(zak): Before pushing to main. See if there is a better way to do this. 
+    // 4 instructions is still a lot cheaper than the non-sse version, but we might be able
+    // to get it cheaper
```
Comment (bvisness) [inline HandmadeMath.h]: @StrangeZak Well, is there? 😃

**Comment by StrangeZak** [comment]
Will fix now.

**Comment by StrangeZak** [comment]
All the comments were old removed them

**Comment by StrangeZak** [comment]
Everything okay now?

**Pull request merged**